Adds KeySpline property to KeyFrame

Deadpikle opened the pull request:
## What does the pull request do?

Adds the potential for `KeySpline` animations to `KeyFrame`. See [docs here](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.animation.keyspline?view=netcore-3.1).


## What is the current behavior?

Avalonia `KeyFrame` objects do not have a `KeySpline`.


## What is the updated/expected behavior with this PR?

New `KeySpline` property and class available!

## How was the solution implemented (if it's not obvious)?

`KeySpline` was ported from WPF code [here](https://github.com/dotnet/wpf/blob/ae1790531c3b993b56eba8b1f0dd395a3ed7de75/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/Animation/KeySpline.cs). I then added `KeySpline` to `KeyFrame` and `AnimatorKeyFrame` so that `Animator<T>` could use it to animate properties.

To test it out, I ran a quick test with my progress ring library/control. Here is the (low resolution GIF) result:

![](https://user-images.githubusercontent.com/5092399/80619592-e688f080-8a12-11ea-8fbd-b6d71c15b70f.gif)

## Notes

* ~~When applying the `KeySpline`, should the `firstKeyFrame` or `lastKeyFrame` `KeySpline` be used in `Animator.InterpolationHandler`? 🤔 Both? Neither?~~ `lastKeyFrame` should be used; see [discussion](https://github.com/AvaloniaUI/Avalonia/pull/3844#discussion_r417994488)
* Since `Avalonia.Point` is in `Avalonia.Visuals`, which isn't used in `Avalonia.Animation`, I changed the use of two `Point` objects in the WPF code to 4 `double` values in the Avalonia code.
* [Reference for where to use `KeySpline` class in frame](https://github.com/dotnet/wpf/blob/ae1790531c3b993b56eba8b1f0dd395a3ed7de75/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/Animation/Generated/SplineKeyFrames.cs#L512)

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes

None.

Deadpikle committed:
Add KeySpline class from WPF source

Deadpikle committed:
Use KeySpline in Animation

Deadpikle committed:
Add some unit tests for KeySpline

Deadpikle committed:
Update documentation

Deadpikle committed:
Update docs again (missed KeySplineTypeConverter)

Gillibald commented:
I think we should reference Avalonia.Visuals  to be able to use Point

Deadpikle commented:
@Gillibald The problem is that `Avalonia.Visuals` references `Avalonia.Animation`. Referencing visual from animation would cause a circular reference, and VS doesn't like that.

Gillibald commented:
We might want to move the animation related members from Avalonia.Visuals to the Avalonia.Animation project if possible. That should solve this.

Gillibald commented:
Okay looks like that is impossible because Visual inherits from StyledElement and that itself inherits Animatable. So the only way I see to solve is merging both projects. Looks like the structure isn't ideal.

Gillibald commented:
Avalonia's Animatable isn't the same as WPF's Animatable. Same thing with Visual.

Deadpikle commented:
@Gillibald Hm, I see. Do you want me to attempt to merge the projects? Such a decision and action is definitely above my proverbial pay grade without someone saying so definitively. 😅 (Maintainers do have access to edit.)

Gillibald commented:
Such a change needs further discussion and needs to be addressed in a different PR

Deadpikle committed:
Merge branch 'master' into feature/keyspline

jmacato commented:
Hmm... i dont understand how Key splines should apply... can you enlighten me @Deadpikle ? First off, does the spline apply in-between keyframes? If so then does it apply in between the n-1th to n keyframe or does it apply n to n+1 keyframes?

jmacato commented on src/Avalonia.Animation/Animators/Animator`1.cs:
```diff
@@ -89,6 +89,9 @@ protected T InterpolationHandler(double animationTime, T neutralValue)
                 newValue = (T)lastKeyframe.Value;
 
+            if (lastKeyframe.KeySpline != null) // TODO: do we use firstKeyFrame or lastKeyframe?!
+                progress = lastKeyframe.KeySpline.GetSplineProgress(progress);
```
This is correct, lastKeyframe should be used for the spline interpolation

jmacato replied:
@Deadpikle  can you confirm that it matches WPF's behavior?

Deadpikle replied:
OK, removed the TODO. Thanks for having me dig more for that. :)

jmacato replied:
> P.S. Noticed that WPF allows for multiple KeyFrame on the same time. Avalonia probably needs to handle that, too.

Well, if the keyframes with the same Cue/KeyTime targets different properties then it should work. It doesnt make sense to allow 2 keyframes targeting the same prop with the same Cue/KeyTime imo.

grokys commented:
Regarding merging projects: the proposal would be to move `Avalonia.Animation` into `Avalonia.Visuals` and make `Animatable` inherit from `StyledElement`, is that correct?

I'm not sure that would work because e.g. `Brush` inherits from `Animatable` and that shouldn't be a `StyledElement` as it's not part of the visual tree.

A better solution might be to move the `Point`/`Size`/`Rect` etc primitives into `Avalonia.Base`?

Gillibald commented:
The issue is here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Visual.cs#L25
>/// Base class for controls that provides rendering and related visual properties.
Visual is a Control?

grokys commented:
@Gillibald see https://gitter.im/AvaloniaUI/avalonia-core?at=5db5c0fa7477946baddda29a for a discussion about this:

```
Steven Kirk @grokys Oct 27 2019 17:08
someone pointed out in the main chat that Visual now implements ILogical after AvaloniaUI/Avalonia#1594 was merged. does this make sense? it came as a surprise to me, as a "visual" was originally a lower-level concept than a "logical" (as it is in WPF)

Jeremy Koritzinsky @jkoritzinsky Oct 27 2019 17:09
I think it makes sense
The layering we have today is all of the data-model layers (base, animation, styling/logical) followed by the rendering layers on top

Steven Kirk @grokys Oct 27 2019 17:16
hmm actually, thinking some more about it, i guess it does make sense in that when we implement our equivalent of FrameworkContentElement etc they can derive from StyledElement but not Visual
````

Gillibald commented:
>A better solution might be to move the Point/Size/Rect etc primitives into Avalonia.Base?

That's the better solution

Deadpikle committed:
Removed TODO now that behavior is verified with WPF

Deadpikle committed:
Merge branch 'feature/keyspline' of https://github.com/Deadpikle/Avalonia into feature/keyspline

Deadpikle commented:
😅 Who knew that using some `double` values instead of some `Point` objects could cause so much discussion!

I'll sit tight on this until I hear for more to do/change. Moving some of those basic classes to Avalonia.Base makes sense to me, an outsider. (If you want me to take a stab and submit a PR for that, let me know.)

Thanks all.

Deadpikle committed:
Merge branch 'master' into feature/keyspline

jmacato commented:
About moving those primitives, I'd like to handle that in a PR. This PR looks good to me :+1:

jmacato reviewed (approved):
About moving those primitives, I'd like to handle that in a PR. This PR looks good to me :+1:

jmacato commented:
Though @Deadpikle it'd be great if you can add unit test for this

Deadpikle committed:
Added a test for KeySpline animation

Deadpikle committed:
Merge branch 'feature/keyspline' of https://github.com/Deadpikle/Avalonia into feature/keyspline

Deadpikle committed:
Merge branch 'master' into feature/keyspline

Deadpikle commented:
@jmacato Alright, I added a unit test for the actual animation process.

Now you're asking, _how did you actually get those crazy values?!_ No, I didn't sit down and math it out. Too lazy for that. 😛 

[WPF Samples has a great UI for viewing KeySpline animations here](https://github.com/microsoft/WPF-Samples/tree/master/Animation/KeySplineAnimations). I took that sample and modified it so that it looked like this:

![WPF-sample](https://user-images.githubusercontent.com/5092399/80752880-545c1780-8afa-11ea-969a-33cf0aaac317.png)

Basically, I added a way to capture the data. (The red bar there and the # were me experimenting.) Now, because CPUs hiccup and getting the time and doubles are inexact and all that, I kept a pretty high tolerance of the animation value checking. But, the values are fairly close, and a user probably won't notice a difference, so I believe the test is sufficient.

Thanks!

> About moving those primitives, I'd like to handle that in a PR.

I interpreted this as "jmacato wants to do this and Deadpikle should not". If I was wrong, let me know. 😅

Deadpikle committed:
Add some notes on how I got the KeySpline test values

Deadpikle committed:
Merge branch 'feature/keyspline' of https://github.com/Deadpikle/Avalonia into feature/keyspline

jmacato committed:
Merge branch 'master' into feature/keyspline

jmacato commented:
> I interpreted this as "jmacato wants to do this and Deadpikle should not". If I was wrong, let me know.

yep that's correct :laughing:
